### ISSUE TITLE
Plan

### ISSUE THREAD

**Issue by ankane** [ankane/torch.rb#1]
Next

- [ ] add `num_workers` option to data loader
- [ ] add `tensor.storage` method
- [ ] add support for loading saved optimizers (`load_state_dict`)

Performance

- [ ] Disable global VM lock for C++ functions

Simplicity

- [ ] Try using `Declarations.yaml` instead of `native_functions.yaml`

Other

- [ ] decide on name (Torch-rb, Torch.rb, Torchrb, Torch Ruby, other suggestions?)
- [ ] test on Windows

Hold

- [ ] Numo bridge for `from_numo` method - `numo_bridge` branch (need to test Numo views and more)

Blocked

- [ ] Numo bridge - see https://github.com/ruby-numo/numo-narray/issues/27
  - [ ] `numo` method - `numo_bridge_to` branch (requires `nary_set_pointer` to be exported from Numo) - https://github.com/ruby-numo/numo-narray/pull/170

Maybe

- [ ] support positional or keyword arguments to make code easier to port - `positional` branch
- [ ] flatten namespace - `Torch::NN::F.relu` -> `Torch.relu`, `Torch::NN::Linear` -> `Torch::Linear`, etc
- [ ] use `autoload` instead of `require`
- [ ] move more logic to C++ frontend (less flexibility but easier maintenance)

Later

- [ ] `torch.hub` - would need to reimplement models in Ruby (similar to TorchVision)
- [ ] run TorchScript models https://pytorch.org/tutorials/advanced/cpp_export.html
- [ ] support ONNX export (not supported in LibTorch) 

Decided against

- Return tensor for comparison operators (`==`, `>`, `>=`, `<`, `<=`). Unlike Python, this causes `Torch.tensor(0) == 1` to return a truthy value, and there's no way to override truthy in Ruby.

**Comment by joaompereira** [comment]
Hi Ankane,
This is a really cool project! Is there at the moment any way to load a previous trained model from python (even if in a more complicated manne than using torch.save/torch.load)?

**Comment by ankane** [comment]
@joaompereira Right now, the best way is to convert the model to ONNX and use [this gem](https://github.com/ankane/onnxruntime).

**Comment by joaompereira** [comment]
@ankane Thanks, that seems a very good option! Would this work with user defined modules?

**Comment by ankane** [comment]
Yeah, afaik

**Comment by ankane** [comment]
fwiw, the latest release now supports [saving and loadings models](https://github.com/ankane/torch.rb#saving-and-loading-models). It uses the same format as PyTorch, so models can be saved in one language and read in another :tada: When saving in Python, use `_use_new_zipfile_serialization=True` until it becomes the default.

**Comment by joaompereira** [comment]
Thanks, that sounds great, I'll check it out!

**Comment by shkit** [comment]
Hi, thanks for very interesting project.

I've tried to compile on Windows with RubyInstaller, but failed due to C++ ABI incompatibility.
(pre-compiled libtorch is compiled by MSVC, but Ruby is compiled by mingw (GCC))

Do you have any information in this area?

I'll try to build Ruby by MSVC, first step.

**Comment by ankane** [comment]
Hey @shkit, I tried compiling it on AppVeyor (https://ci.appveyor.com/project/ankane/torch-rb), but believe it failed due to the issue you mentioned above.

I get the impression RubyInstaller is the most popular way to install Ruby on Windows, so it may make sense to try and compile Libtorch with mingw, so users don't need to change their Ruby.

**Comment by shkit** [comment]
Hi @ankane,
I agree with you, but supporting CUDA compiler (NVCC) requires MSVC.
  https://discuss.pytorch.org/t/can-i-build-libtorch-from-source-with-mingw/52632

This is one of the troublesome thing on Windows environment...

**Comment by ankane** [comment]
Good catch. Since a separate Ruby is essentially needed, it may be easiest to just run the Docker image. https://github.com/ankane/ml-stack

**Comment by slowpilot** [comment]
Hi, first of all, let me thank you for starting such great work. Ruby, being a great language, nevertheless, is starving for ml/ds gems. I would like to contribute somehow on building ruby torch. I own a bunch of multi-GPU machines (4x, 6x and 14x GPU builds that I use for ML projects), so I may be useful for e.g. testing in such environment or ruby ditstributed data parallel implementation (NCCL, GLOO, MPI).

**Comment by ankane** [comment]
Hey @slowpilot, thanks for the interest in helping out. There are some ideas in the first comment or you could start using the library and contributing as things come up. I haven't done any work on distributed processing yet, so am not sure what that looks like.

**Comment by SamurotTt** [comment]
Ohayo @ankane, I've A LOT INTEREST in helping you in this project, but I don't have the necessary knowledge, I mean, Im free to contribute to the simple tasks to achive more learnings about complex concepts in programming, so if you need an assistant please contact me 'cause I need someone to instruct me on the path of Ruby.
